File: lab/backend/src/vm_lab/routers/measure.py
"""POST /api/measure — teach → find → measure → judge, the last step.

Places a `MetrologyModel` at a fixture pose (explicit, or the top auto-found match) and
measures each requested object. Everything the frontend needs to draw is computed here,
in source-image pixel coordinates — the client never reconstructs geometry (see
lab/README.md's "source-frame" rule).

Two passes over the same calipers, deliberately:
1. `vm.MetrologyModel.apply` does the real measurement — robust fit, residuals.
2. This module's own `_place_calipers` (mirroring `MetrologyModel::measure_one`,
   `crates/vision-metrology/src/measure/model.rs`) re-derives each caliper and measures
   it again, only to report **why** a caliper was rejected and to expose its raw profile
   for `LineProfile` — detail `apply`'s binding does not surface. Both passes share the
   same `MeasureConfig`, so they agree on every caliper that succeeds.
"""

from __future__ import annotations

import math
from typing import Any

import numpy as np
from fastapi import APIRouter, HTTPException

import vision_metrology as vm

from vm_lab.geometry import perp, to_scene
from vm_lab.routers.find import run_find
from vm_lab.schemas import (
    CaliperProfileOut,
    CaliperResultOut,
    EdgeMarkOut,
    FindRequest,
    FixtureIn,
    MeasureObjectIn,
    MeasureObjectResultOut,
    MeasureRequest,
    MeasureResponse,
    OverlayPrimitiveOut,
)
from vm_lab.store import store

router = APIRouter(prefix="/api/measure", tags=["measure"])


def _resolve_metric(req: MeasureRequest) -> tuple[vm.CameraModel, np.ndarray, vm.Plane3] | None:
    """`(camera, pose, plane)` when `req.calibration_id` is set, else `None` — the
    off switch for every mm computation below."""
    if req.calibration_id is None:
        return None
    if req.calibration_id not in store.calibrations:
        raise HTTPException(404, f"no such calibration: {req.calibration_id}")
    cameras = store.get_calibration(req.calibration_id)
    if not 0 <= req.camera_index < len(cameras):
        raise HTTPException(
            400, f"camera_index {req.camera_index} out of range (calibration has {len(cameras)} cameras)"
        )
    camera, pose = cameras[req.camera_index]
    plane = vm.Plane3((req.plane.nx, req.plane.ny, req.plane.nz), req.plane.d)
    return camera, pose, plane


def _pixel_to_plane_mm(metric: tuple[vm.CameraModel, np.ndarray, vm.Plane3], x: float, y: float) -> tuple[float, float] | None:
    camera, pose, plane = metric
    pixels = np.array([[x, y]], dtype=np.float32)
    out = vm.pixel_to_plane(camera, pose, plane, pixels)[0]
    if not np.isfinite(out).all():
        return None
    return float(out[0]), float(out[1])


def _measure_config(obj: MeasureObjectIn) -> vm.MeasureConfig:
    # `select="strongest"` matches `MetrologyObject::new`'s Rust default (model.rs) —
    # forced explicitly here rather than left to the binding's own default, so this
    # config is guaranteed identical to the one handed to `vm.MetrologyObject` below.
    return vm.MeasureConfig(
        sigma=obj.measure.sigma,
        threshold=obj.measure.threshold,
        polarity=obj.measure.polarity,
        select="strongest",
        max_obliquity_deg=obj.measure.max_obliquity_deg,
    )


def _fit_config(obj: MeasureObjectIn) -> vm.FitConfig:
    return vm.FitConfig(loss=obj.fit.loss, inlier_tol=obj.fit.inlier_tol)


def _vm_shape(obj: MeasureObjectIn) -> Any:
    if obj.kind == "circle":
        if obj.cx is None or obj.cy is None or obj.r is None:
            raise HTTPException(400, "circle object needs cx, cy, r")
        arc = None
        if obj.arc is not None:
            arc = (math.radians(obj.arc[0]), math.radians(obj.arc[1]))
        return vm.MetrologyShape.circle((obj.cx, obj.cy), obj.r, arc)
    if obj.ax is None or obj.ay is None or obj.bx is None or obj.by is None:
        raise HTTPException(400, "line object needs ax, ay, bx, by")
    return vm.MetrologyShape.line((obj.ax, obj.ay), (obj.bx, obj.by))


def _place_calipers(
    obj: MeasureObjectIn, origin: tuple[float, float], x: float, y: float, angle: float, scale: float
) -> list[tuple[str, dict[str, Any]]]:
    """Re-derives each caliper's placement, mirroring `MetrologyModel::measure_one`."""
    n = obj.n_calipers
    specs: list[tuple[str, dict[str, Any]]] = []
    half_len = obj.caliper_len * scale
    half_width = obj.caliper_width * scale
    if obj.kind == "line":
        assert obj.ax is not None and obj.ay is not None and obj.bx is not None and obj.by is not None
        ax, ay = to_scene(obj.ax, obj.ay, origin, x, y, angle, scale)
        bx, by = to_scene(obj.bx, obj.by, origin, x, y, angle, scale)
        length = math.hypot(bx - ax, by - ay)
        if length < 1e-6:
            raise HTTPException(400, "metrology line has zero length")
        alx, aly = (bx - ax) / length, (by - ay) / length
        nx, ny = perp(alx, aly)
        cal_angle = math.atan2(ny, nx)
        for i in range(n):
            f = i / (n - 1)
            cx, cy = ax + (bx - ax) * f, ay + (by - ay) * f
            specs.append(("rect", dict(center=(cx, cy), angle=cal_angle, half_len=half_len, half_width=half_width)))
    else:
        assert obj.cx is not None and obj.cy is not None and obj.r is not None
        cx, cy = to_scene(obj.cx, obj.cy, origin, x, y, angle, scale)
        radius = obj.r * scale
        if obj.arc is not None:
            start, extent = math.radians(obj.arc[0]), math.radians(obj.arc[1])
        else:
            start, extent = 0.0, 2 * math.pi
        start = start + angle
        closed = abs(abs(extent) - 2 * math.pi) < 1e-3
        denom = n if closed else (n - 1)
        for i in range(n):
            phi = start + extent * i / denom
            specs.append(
                ("radial", dict(center=(cx, cy), radius=radius, angle=phi, half_len=half_len, half_width=half_width))
            )
    return specs


def _measure_calipers(
    specs: list[tuple[str, dict[str, Any]]],
    config: vm.MeasureConfig,
    img: Any,
    metric: tuple[vm.CameraModel, np.ndarray, vm.Plane3] | None = None,
) -> tuple[list[CaliperResultOut], list[OverlayPrimitiveOut]]:
    results: list[CaliperResultOut] = []
    overlay: list[OverlayPrimitiveOut] = []
    step_px = config.step
    for i, (kind, kwargs) in enumerate(specs):
        cal = vm.Caliper.rect(kwargs["center"], kwargs["angle"], kwargs["half_len"], kwargs["half_width"], config) \
            if kind == "rect" else \
            vm.Caliper.radial(kwargs["center"], kwargs["radius"], kwargs["angle"], kwargs["half_len"], kwargs["half_width"], config)
        cx, cy = kwargs["center"]
        box_angle = kwargs["angle"]
        try:
            edges = cal.measure(img)
        except vm.MeasureRejected as exc:
            reason = str(exc.args[0]) if exc.args else "rejected"
            profile = CaliperProfileOut(values=list(cal.profile()), step_px=step_px, edges=[])
            results.append(CaliperResultOut(index=i, status="rejected", reason=reason, profile=profile))
            overlay.append(
                OverlayPrimitiveOut(
                    kind="caliper", tone="defect", cx=cx, cy=cy,
                    width=2 * kwargs["half_len"], height=2 * kwargs["half_width"], angle=box_angle,
                )
            )
            continue
        edge = edges[0]
        mm = _pixel_to_plane_mm(metric, edge.x, edge.y) if metric is not None else None
        profile = CaliperProfileOut(
            values=list(cal.profile()), step_px=step_px,
            edges=[
                EdgeMarkOut(
                    pos_px=edge.t,
                    polarity=edge.polarity,
                    x_mm=mm[0] if mm is not None else None,
                    y_mm=mm[1] if mm is not None else None,
                )
            ],
        )
        results.append(CaliperResultOut(index=i, status="hit", profile=profile))
        overlay.append(
            OverlayPrimitiveOut(
                kind="caliper", tone="signal", cx=cx, cy=cy,
                width=2 * kwargs["half_len"], height=2 * kwargs["half_width"], angle=box_angle,
            )
        )
        overlay.append(OverlayPrimitiveOut(kind="point", tone="signal", x=edge.x, y=edge.y, cross=True))
    return results, overlay


def _resolve_fixture(req: MeasureRequest) -> tuple[FixtureIn, str]:
    if req.fixture is not None:
        return req.fixture, "explicit"
    find_req = FindRequest(image_id=req.image_id, model_id=req.model_id, min_score=req.min_score, max_matches=1)
    matches = run_find(find_req)
    if not matches:
        raise HTTPException(422, "auto-find found no match at or above min_score")
    best = max(matches, key=lambda m: m.score)
    return FixtureIn(x=best.x, y=best.y, angle=best.angle, scale=best.scale), "auto_find"


@router.post("", response_model=MeasureResponse)
async def measure(req: MeasureRequest) -> MeasureResponse:
    if req.image_id not in store.images:
        raise HTTPException(404, f"no such image: {req.image_id}")
    if req.model_id not in store.models:
        raise HTTPException(404, f"no such model: {req.model_id}")
    if not req.objects:
        raise HTTPException(400, "at least one object is required")

    fixture, source = _resolve_fixture(req)
    img = store.load_array(req.image_id)
    model = store.get_model(req.model_id)
    origin = model.origin
    metric = _resolve_metric(req)

    metrology_model = vm.MetrologyModel()
    for obj in req.objects:
        if obj.n_calipers < 2:
            raise HTTPException(400, "n_calipers must be >= 2")
        metrology_model.add(
            vm.MetrologyObject(
                shape=_vm_shape(obj),
                n_calipers=obj.n_calipers,
                caliper_len=obj.caliper_len,
                caliper_width=obj.caliper_width,
                measure=_measure_config(obj),
                fit=_fit_config(obj),
            )
        )

    raw_results = metrology_model.apply(
        img, x=fixture.x, y=fixture.y, angle=fixture.angle, scale=fixture.scale, origin=origin
    )

    out_objects: list[MeasureObjectResultOut] = []
    for obj, raw in zip(req.objects, raw_results, strict=True):
        specs = _place_calipers(obj, origin, fixture.x, fixture.y, fixture.angle, fixture.scale)
        config = _measure_config(obj)
        calipers, cal_overlay = _measure_calipers(specs, config, img, metric)

        if isinstance(raw, vm.MetrologyError):
            out_objects.append(
                MeasureObjectResultOut(kind="error", label=obj.label, message=raw.message, calipers=calipers, overlay=cal_overlay)
            )
            continue

        overlay = list(cal_overlay)
        if raw.kind == "circle" and raw.circle is not None:
            overlay.append(OverlayPrimitiveOut(kind="circle", tone="normal", cx=raw.circle.cx, cy=raw.circle.cy, r=raw.circle.r))
        elif raw.kind == "line" and raw.line is not None:
            line = raw.line
            ts = [(e.x - line.px) * line.dx + (e.y - line.py) * line.dy for e in raw.hits]
            tmin, tmax = (min(ts), max(ts)) if ts else (-obj.caliper_len, obj.caliper_len)
            overlay.append(
                OverlayPrimitiveOut(
                    kind="segment", tone="normal",
                    x1=line.px + line.dx * tmin, y1=line.py + line.dy * tmin,
                    x2=line.px + line.dx * tmax, y2=line.py + line.dy * tmax,
                )
            )

        circle_cx_mm = circle_cy_mm = circle_r_mm = None
        if metric is not None and raw.kind == "circle" and raw.circle is not None:
            center_mm = _pixel_to_plane_mm(metric, raw.circle.cx, raw.circle.cy)
            # Radius via two diametral points (exact per point; the distance
            # between them is a fronto-parallel-view approximation of the
            # radius under a tilted camera, see MeasureObjectResultOut docs).
            p1_mm = _pixel_to_plane_mm(metric, raw.circle.cx + raw.circle.r, raw.circle.cy)
            p2_mm = _pixel_to_plane_mm(metric, raw.circle.cx - raw.circle.r, raw.circle.cy)
            if center_mm is not None:
                circle_cx_mm, circle_cy_mm = center_mm
            if p1_mm is not None and p2_mm is not None:
                circle_r_mm = math.hypot(p1_mm[0] - p2_mm[0], p1_mm[1] - p2_mm[1]) / 2.0

        out_objects.append(
            MeasureObjectResultOut(
                kind=raw.kind,
                label=obj.label,
                circle_cx=raw.circle.cx if raw.circle else None,
                circle_cy=raw.circle.cy if raw.circle else None,
                circle_r=raw.circle.r if raw.circle else None,
                line_px=raw.line.px if raw.line else None,
                line_py=raw.line.py if raw.line else None,
                line_dx=raw.line.dx if raw.line else None,
                line_dy=raw.line.dy if raw.line else None,
                rms=raw.rms,
                max_dev=raw.max_dev,
                n_used=raw.n_used,
                circle_cx_mm=circle_cx_mm,
                circle_cy_mm=circle_cy_mm,
                circle_r_mm=circle_r_mm,
                calipers=calipers,
                overlay=overlay,
            )
        )

    return MeasureResponse(fixture=fixture, fixture_source=source, objects=out_objects)

File: lab/frontend/src-tauri/src/commands/measure.rs
//! `measure` — mirrors `lab/backend/src/vm_lab/routers/measure.py`.
//!
//! Two passes over the same calipers, same reasoning as the Python router: (1)
//! `MetrologyModel::apply` does the real measurement (robust fit, residuals); (2) this
//! module re-measures each caliper individually, using the *exact same placement*
//! `apply` used internally (`measure::diagnostics::layout` — the W6 Part A API), only to
//! report which caliper was rejected and why, and to expose its raw profile. Both passes
//! share the same `MeasureConfig`, so they agree on every caliper that succeeds.

use std::collections::HashMap;

use vision_metrology::fit::{FitConfig, RobustLoss};
use vision_metrology::measure::diagnostics::{CaliperShape, layout};
use vision_metrology::measure::{
    Caliper, EdgeSelect, MeasureConfig as NativeMeasureConfig, MetrologyFit, MetrologyModel,
    MetrologyObject, MetrologyShape, PolaritySelect, RejectReason,
};
use vision_metrology::metric::{CameraModel, Plane3, Pose3, pixel_to_plane};
use vm_primitives::{Point2f, Similarity2f, Vec2f, similarity_from_parts, wrap_angle};

use super::find::run_find;
use crate::error::{AppError, AppResult, not_found};
use crate::state::AppState;
use crate::types::{
    CaliperProfileOut, CaliperResultOut, EdgeMarkOut, FindRequest, FixtureIn, MeasureObjectIn,
    MeasureObjectResultOut, MeasureRequest, MeasureResponse, MeasureShapeKind, OverlayPrimitiveOut,
};

/// `Translation(position) ∘ sR ∘ Translation(−origin)`, as a similarity — identical to
/// `ShapeMatch::pose`'s own construction (mirrors `vm-python`'s `measure_py::pose_from`).
fn pose_from(position: Point2f, angle: f32, scale: f32, origin: Point2f) -> Similarity2f {
    let (sn, cs) = wrap_angle(angle).sin_cos();
    let t = Vec2f::new(
        position.x - scale * (cs * origin.x - sn * origin.y),
        position.y - scale * (sn * origin.x + cs * origin.y),
    );
    similarity_from_parts(t, wrap_angle(angle), scale)
}

fn polarity_from(s: Option<&str>) -> PolaritySelect {
    match s {
        Some("rising") | Some("dark_to_bright") => PolaritySelect::Rising,
        Some("falling") | Some("bright_to_dark") => PolaritySelect::Falling,
        _ => PolaritySelect::Any,
    }
}

/// `inlier_tol` doubles as the robust-loss radius here — the lab's own `FitConfigIn`
/// only exposes `loss`/`inlier_tol`, no separate loss-scale knob, and this is the more
/// useful reading of the pair for a caller who never touches RANSAC (which needs
/// `ransac_iters` set too, and this command exposes no such field).
fn fit_from(fit: Option<&crate::types::FitConfigIn>) -> FitConfig {
    let loss_str = fit.and_then(|f| f.loss.as_deref()).unwrap_or("l2");
    let radius = fit.and_then(|f| f.inlier_tol).unwrap_or(2.0);
    let loss = match loss_str {
        "huber" => RobustLoss::Huber { k: radius },
        "tukey" => RobustLoss::Tukey { c: radius },
        _ => RobustLoss::None,
    };
    FitConfig {
        loss,
        ..FitConfig::default()
    }
}

fn measure_config_from(m: Option<&crate::types::MeasureConfigIn>) -> NativeMeasureConfig {
    let base = NativeMeasureConfig {
        select: EdgeSelect::Strongest,
        ..NativeMeasureConfig::default()
    };
    NativeMeasureConfig {
        sigma: m.and_then(|c| c.sigma).unwrap_or(base.sigma),
        threshold: m.and_then(|c| c.threshold).unwrap_or(base.threshold),
        polarity: m
            .and_then(|c| c.polarity.as_deref())
            .map(|s| polarity_from(Some(s)))
            .unwrap_or(base.polarity),
        max_obliquity_deg: m
            .and_then(|c| c.max_obliquity_deg)
            .unwrap_or(base.max_obliquity_deg),
        ..base
    }
}

fn shape_from(obj: &MeasureObjectIn) -> AppResult<MetrologyShape> {
    match obj.kind {
        MeasureShapeKind::Circle => {
            let (cx, cy, r) = (
                obj.cx
                    .ok_or_else(|| AppError("circle object needs cx".into()))?,
                obj.cy
                    .ok_or_else(|| AppError("circle object needs cy".into()))?,
                obj.r
                    .ok_or_else(|| AppError("circle object needs r".into()))?,
            );
            let arc = obj.arc.map(|(a, b)| (a.to_radians(), b.to_radians()));
            Ok(MetrologyShape::Circle {
                center: Point2f::new(cx, cy),
                radius: r,
                arc,
            })
        }
        MeasureShapeKind::Line => {
            let (ax, ay, bx, by) = (
                obj.ax
                    .ok_or_else(|| AppError("line object needs ax".into()))?,
                obj.ay
                    .ok_or_else(|| AppError("line object needs ay".into()))?,
                obj.bx
                    .ok_or_else(|| AppError("line object needs bx".into()))?,
                obj.by
                    .ok_or_else(|| AppError("line object needs by".into()))?,
            );
            Ok(MetrologyShape::Line {
                a: Point2f::new(ax, ay),
                b: Point2f::new(bx, by),
            })
        }
    }
}

type Metric = (CameraModel, Pose3, Plane3);

fn resolve_metric(state: &AppState, req: &MeasureRequest) -> AppResult<Option<Metric>> {
    let Some(cal_id) = &req.calibration_id else {
        return Ok(None);
    };
    let calibrations = state
        .calibrations
        .lock()
        .expect("calibrations mutex poisoned");
    let entry = calibrations
        .get(cal_id)
        .ok_or_else(|| not_found("calibration", cal_id))?;
    let (camera, pose) = *entry.cameras.get(req.camera_index).ok_or_else(|| {
        AppError(format!(
            "camera_index {} out of range (calibration has {} cameras)",
            req.camera_index,
            entry.cameras.len()
        ))
    })?;
    let plane = Plane3 {
        n: vm_primitives::Vec3f::new(req.plane.nx, req.plane.ny, req.plane.nz),
        d: req.plane.d,
    };
    Ok(Some((camera, pose, plane)))
}

fn pixel_to_plane_mm(metric: &Metric, p: Point2f) -> Option<(f32, f32)> {
    let (camera, pose, plane) = metric;
    pixel_to_plane(camera, pose, plane, p).map(|q| (q.x, q.y))
}

fn resolve_fixture(state: &AppState, req: &MeasureRequest) -> AppResult<(FixtureIn, &'static str)> {
    if let Some(f) = req.fixture {
        return Ok((f, "explicit"));
    }
    let find_req = FindRequest {
        image_id: req.image_id.clone(),
        model_id: req.model_id.clone(),
        min_score: req.min_score,
        max_matches: Some(1),
        ..FindRequest::default()
    };
    let matches = run_find(state, &find_req)?;
    let best = matches
        .into_iter()
        .max_by(|a, b| a.score.partial_cmp(&b.score).expect("score is never NaN"))
        .ok_or_else(|| AppError("auto-find found no match at or above min_score".into()))?;
    Ok((
        FixtureIn {
            x: best.position.x,
            y: best.position.y,
            angle: best.angle(),
            scale: best.scale(),
        },
        "auto_find",
    ))
}

fn caliper_for(shape: &CaliperShape, config: NativeMeasureConfig) -> Caliper {
    match *shape {
        CaliperShape::Rect(r) => Caliper::rect(r, config),
        CaliperShape::Radial(r) => Caliper::radial(r, config),
    }
}

fn reject_reason_str(r: RejectReason) -> &'static str {
    match r {
        RejectReason::ProfileTooShort => "profile_too_short",
        RejectReason::NoEdge => "no_edge",
        RejectReason::WrongPolarity => "wrong_polarity",
        RejectReason::TooOblique => "too_oblique",
        RejectReason::OffImage => "off_image",
    }
}

fn placement_geometry(shape: &CaliperShape) -> (Point2f, f32, f32, f32) {
    match *shape {
        CaliperShape::Rect(r) => (r.center, r.angle, r.half_len, r.half_width),
        CaliperShape::Radial(r) => (r.center, r.angle, r.half_len, r.half_width),
    }
}

fn measure_calipers(
    shapes: &[CaliperShape],
    config: NativeMeasureConfig,
    img: &vm_primitives::ImageView<'_, u8>,
    metric: Option<&Metric>,
) -> (Vec<CaliperResultOut>, Vec<OverlayPrimitiveOut>) {
    let mut results = Vec::with_capacity(shapes.len());
    let mut overlay = Vec::new();
    let step_px = config.step;

    for (i, shape) in shapes.iter().enumerate() {
        let mut cal = caliper_for(shape, config);
        let (center, angle, half_len, half_width) = placement_geometry(shape);
        match cal.measure(img) {
            Err(reason) => {
                let profile = CaliperProfileOut {
                    values: cal.profile().to_vec(),
                    step_px,
                    edges: Vec::new(),
                };
                results.push(CaliperResultOut {
                    index: i,
                    status: "rejected",
                    reason: Some(reject_reason_str(reason).to_string()),
                    profile,
                });
                overlay.push(OverlayPrimitiveOut {
                    kind: "caliper",
                    tone: Some("defect"),
                    cx: Some(center.x),
                    cy: Some(center.y),
                    width: Some(2.0 * half_len),
                    height: Some(2.0 * half_width),
                    angle: Some(angle),
                    ..Default::default()
                });
            }
            Ok(edges) => {
                let edge = edges[0];
                let mm = metric.and_then(|m| pixel_to_plane_mm(m, edge.p));
                let profile = CaliperProfileOut {
                    values: cal.profile().to_vec(),
                    step_px,
                    edges: vec![EdgeMarkOut {
                        pos_px: edge.t,
                        polarity: format!("{:?}", edge.polarity).to_lowercase(),
                        x_mm: mm.map(|m| m.0),
                        y_mm: mm.map(|m| m.1),
                    }],
                };
                results.push(CaliperResultOut {
                    index: i,
                    status: "hit",
                    reason: None,
                    profile,
                });
                overlay.push(OverlayPrimitiveOut {
                    kind: "caliper",
                    tone: Some("signal"),
                    cx: Some(center.x),
                    cy: Some(center.y),
                    width: Some(2.0 * half_len),
                    height: Some(2.0 * half_width),
                    angle: Some(angle),
                    ..Default::default()
                });
                overlay.push(OverlayPrimitiveOut {
                    kind: "point",
                    tone: Some("signal"),
                    x: Some(edge.p.x),
                    y: Some(edge.p.y),
                    cross: Some(true),
                    ..Default::default()
                });
            }
        }
    }
    (results, overlay)
}

pub fn measure(state: &AppState, req: MeasureRequest) -> AppResult<MeasureResponse> {
    if req.objects.is_empty() {
        return Err("at least one object is required".into());
    }
    for obj in &req.objects {
        if obj.n_calipers < 2 {
            return Err("n_calipers must be >= 2".into());
        }
    }

    let (fixture, source) = resolve_fixture(state, &req)?;
    let metric = resolve_metric(state, &req)?;

    let origin = {
        let models = state.models.lock().expect("models mutex poisoned");
        models
            .get(&req.model_id)
            .ok_or_else(|| not_found("model", &req.model_id))?
            .model
            .origin()
    };
    // After the models lock is released: `decoded` takes the images lock.
    let image = state.decoded(&req.image_id)?;

    let pose = pose_from(
        Point2f::new(fixture.x, fixture.y),
        fixture.angle,
        fixture.scale,
        origin,
    );

    let mut metrology_model = MetrologyModel::new();
    for obj in &req.objects {
        let shape = shape_from(obj)?;
        metrology_model.add(MetrologyObject {
            shape,
            n_calipers: obj.n_calipers,
            caliper_len: obj.caliper_len,
            caliper_width: obj.caliper_width,
            measure: measure_config_from(obj.measure.as_ref()),
            fit: fit_from(obj.fit.as_ref()),
        });
    }

    let view = image.as_view();
    let raw_results = metrology_model.apply(&view, &pose);
    let placements = layout(&metrology_model, &pose);
    let mut by_object: HashMap<usize, Vec<CaliperShape>> = HashMap::new();
    for p in placements {
        by_object.entry(p.object_index).or_default().push(p.shape);
    }

    let mut out_objects = Vec::with_capacity(req.objects.len());
    for (i, (obj, raw)) in req.objects.iter().zip(raw_results).enumerate() {
        let config = measure_config_from(obj.measure.as_ref());
        let shapes = by_object.remove(&i).unwrap_or_default();
        let (calipers, cal_overlay) = measure_calipers(&shapes, config, &view, metric.as_ref());

        let raw = match raw {
            Err(e) => {
                out_objects.push(MeasureObjectResultOut {
                    kind: "error",
                    label: obj.label.clone(),
                    message: Some(e.to_string()),
                    circle_cx: None,
                    circle_cy: None,
                    circle_r: None,
                    line_px: None,
                    line_py: None,
                    line_dx: None,
                    line_dy: None,
                    rms: None,
                    max_dev: None,
                    n_used: None,
                    circle_cx_mm: None,
                    circle_cy_mm: None,
                    circle_r_mm: None,
                    calipers,
                    overlay: cal_overlay,
                });
                continue;
            }
            Ok(r) => r,
        };

        let mut overlay = cal_overlay;
        let (kind, circle_cx, circle_cy, circle_r, line_px, line_py, line_dx, line_dy) =
            match &raw.fit {
                MetrologyFit::Circle(f) => {
                    overlay.push(OverlayPrimitiveOut {
                        kind: "circle",
                        tone: Some("normal"),
                        cx: Some(f.model.center.x),
                        cy: Some(f.model.center.y),
                        r: Some(f.model.radius),
                        ..Default::default()
                    });
                    (
                        "circle",
                        Some(f.model.center.x),
                        Some(f.model.center.y),
                        Some(f.model.radius),
                        None,
                        None,
                        None,
                        None,
                    )
                }
                MetrologyFit::Line(f) => {
                    let ts: Vec<f32> = raw
                        .hits
                        .iter()
                        .map(|e| {
                            (e.p.x - f.model.p.x) * f.model.dir.x
                                + (e.p.y - f.model.p.y) * f.model.dir.y
                        })
                        .collect();
                    let (tmin, tmax) = if ts.is_empty() {
                        (-obj.caliper_len, obj.caliper_len)
                    } else {
                        (
                            ts.iter().cloned().fold(f32::INFINITY, f32::min),
                            ts.iter().cloned().fold(f32::NEG_INFINITY, f32::max),
                        )
                    };
                    overlay.push(OverlayPrimitiveOut {
                        kind: "segment",
                        tone: Some("normal"),
                        x1: Some(f.model.p.x + f.model.dir.x * tmin),
                        y1: Some(f.model.p.y + f.model.dir.y * tmin),
                        x2: Some(f.model.p.x + f.model.dir.x * tmax),
                        y2: Some(f.model.p.y + f.model.dir.y * tmax),
                        ..Default::default()
                    });
                    (
                        "line",
                        None,
                        None,
                        None,
                        Some(f.model.p.x),
                        Some(f.model.p.y),
                        Some(f.model.dir.x),
                        Some(f.model.dir.y),
                    )
                }
            };

        let (mut circle_cx_mm, mut circle_cy_mm, mut circle_r_mm) = (None, None, None);
        if let (Some(m), MetrologyFit::Circle(f)) = (&metric, &raw.fit) {
            let center_mm = pixel_to_plane_mm(m, f.model.center);
            let p1_mm = pixel_to_plane_mm(
                m,
                Point2f::new(f.model.center.x + f.model.radius, f.model.center.y),
            );
            let p2_mm = pixel_to_plane_mm(
                m,
                Point2f::new(f.model.center.x - f.model.radius, f.model.center.y),
            );
            if let Some((x, y)) = center_mm {
                circle_cx_mm = Some(x);
                circle_cy_mm = Some(y);
            }
            if let (Some(p1), Some(p2)) = (p1_mm, p2_mm) {
                circle_r_mm = Some(((p1.0 - p2.0).powi(2) + (p1.1 - p2.1).powi(2)).sqrt() / 2.0);
            }
        }

        out_objects.push(MeasureObjectResultOut {
            kind,
            label: obj.label.clone(),
            message: None,
            circle_cx,
            circle_cy,
            circle_r,
            line_px,
            line_py,
            line_dx,
            line_dy,
            rms: Some(raw.rms()),
            max_dev: Some(raw.max_dev()),
            n_used: Some(raw.n_used()),
            circle_cx_mm,
            circle_cy_mm,
            circle_r_mm,
            calipers,
            overlay,
        });
    }

    Ok(MeasureResponse {
        fixture,
        fixture_source: source,
        objects: out_objects,
    })
}
